src/stream.jl

```diff
@@ -325,15 +327,11 @@ end
 # --------------
 
 function Base.read(stream::TranscodingStream, ::Type{UInt8})
-    # eof and ready_to_read! are inlined here because ready_to_read! is very slow and eof is broken
-    eof = buffersize(stream.buffer1) == 0
-    state = stream.state
-    mode = state.mode
-    if !(mode == :read || mode == :stop)
-        changemode!(stream, :read)
-    end
-    if eof && sloweof(stream)
-        throw(EOFError())
+    if eof(stream)
+        ready_to_read!(stream)
```

Comment:
This should be implied by !eof
```suggestion
```

Reply:
This change means an `EOFError` is thrown when reading a byte from a closed stream.
This seems to match what happens with `IOStream`
```julia
julia> p, io = mktemp()
("/tmp/jl_GKa3GO", IOStream(<fd 21>))

julia> close(io)

julia> read(io, UInt8)
ERROR: EOFError: read end of file
Stacktrace:
 [1] read(s::IOStream, ::Type{UInt8})
   @ Base ./iostream.jl:400
 [2] top-level scope
   @ REPL[43]:1
```

But doesn't match with what `IOBuffer` does:
```julia
julia> stream = IOBuffer();

julia> close(stream)

julia> read(stream, UInt8)
ERROR: ArgumentError: read failed, IOBuffer is not readable
Stacktrace:
 [1] _throw_not_readable()
   @ Base ./iobuffer.jl:165
 [2] read(from::IOBuffer, ::Type{UInt8})
   @ Base ./iobuffer.jl:218
 [3] top-level scope
   @ REPL[49]:1
```

Reply:
True, though I think `eof` could optionally throw a state-exception in that case also sometimes. The ambiguity occurs when a stream is ambiguously readable or writable, which sometimes cannot be avoided. So seekable files usually `close` both ends at once (and eof is a transient condition, which changes to an error after calling close), whereas (non-seekable) streams the eof is a property of reading and isopen is a property of writing

Reply:
I'm undoing the change because I like how `IOBuffer` is handling this better than `IOStream`.

For example, I think `read(io, String)` after close should error instead of returning an empty string:
```julia
julia> p, io = mktemp()

julia> close(io)

julia> read(io, String)
""
``` 
```julia
julia> stream = IOBuffer();

julia> close(stream)

julia> read(stream, String)
ERROR: ArgumentError: read failed, IOBuffer is not readable

```

Reply:
Yes, but that is already eof’s job to throw when there is a state violation, so it should not be necessary for read to also redundantly check

Reply:
This would break this test that was added in #32:

https://github.com/JuliaIO/TranscodingStreams.jl/blob/92adcd82f87562bd350829195c5d3ba32fe3f9fa/test/codecnoop.jl#L283-L287

Reply:
I agree that eof should throw if the stream is not in a `readable` state. 

However this seems like a breaking change, so I will split this PR into smaller parts, and add the eof changes in v0.11 for v0.11 I would also like to fix #93, #109 and remove the `seekend` implementation.